packages/in_app_purchase/in_app_purchase_ios/lib/src/in_app_purchase_ios_platform.dart

```diff
@@ -0,0 +1,223 @@
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:flutter/foundation.dart';
+import 'package:flutter/services.dart';
+import 'package:in_app_purchase_ios/src/in_app_purchase_ios_platform_addition.dart';
+import 'package:in_app_purchase_platform_interface/in_app_purchase_platform_interface.dart';
+
+import '../in_app_purchase_ios.dart';
+import '../store_kit_wrappers.dart';
+
+/// [IAPError.code] code for failed purchases.
+final String kPurchaseErrorCode = 'purchase_error';
+
+/// [IAPError.code] code used when a query for previouys transaction has failed.
+final String kRestoredPurchaseErrorCode = 'restore_transactions_failed';
+
+/// [IAPError.code] code used when a consuming a purchased item fails.
+final String kConsumptionFailedErrorCode = 'consume_purchase_failed';
+
+/// Indicates store front is Apple AppStore.
+final String kIAPSource = 'app_store';
+
+/// An [InAppPurchasePlatform] that wraps StoreKit.
+///
+/// This translates various `StoreKit` calls and responses into the
+/// generic plugin API.
+class InAppPurchaseIosPlatform extends InAppPurchasePlatform {
+  /// Returns the singleton instance of the [InAppPurchaseIosPlatform] that should be
+  /// used across the app.
+  static InAppPurchaseIosPlatform get instance => _getOrCreateInstance();
+  static InAppPurchaseIosPlatform? _instance;
+  static late SKPaymentQueueWrapper _skPaymentQueueWrapper;
+  static late _TransactionObserver _observer;
+
+  /// Creates an [InAppPurchaseIosPlatform] object.
+  ///
+  /// This constructor should only be used for testing, for any other purpose
+  /// get the connection from the [instance] getter.
+  @visibleForTesting
+  InAppPurchaseIosPlatform();
+
+  Stream<List<PurchaseDetails>> get purchaseStream =>
+      _observer.purchaseUpdatedController.stream;
+
+  /// Callback handler for transaction status changes.
+  @visibleForTesting
+  static SKTransactionObserverWrapper get observer => _observer;
+
+  static InAppPurchaseIosPlatform _getOrCreateInstance() {
+    if (_instance != null) {
+      return _instance!;
+    }
+
+    // Register the [InAppPurchaseIosPlatformAddition] containing iOS
+    // platform-specific functionality.
+    InAppPurchasePlatformAddition.instance = InAppPurchaseIosPlatformAddition();
+
+    // Register the platform-specific implementation of the idiomatic
+    // InAppPurchase API.
+    _instance = InAppPurchaseIosPlatform();
+    _skPaymentQueueWrapper = SKPaymentQueueWrapper();
+    _observer = _TransactionObserver(StreamController.broadcast());
+    _skPaymentQueueWrapper.setTransactionObserver(observer);
+    return _instance!;
+  }
+
+  @override
+  Future<bool> isAvailable() => SKPaymentQueueWrapper.canMakePayments();
+
+  @override
+  Future<bool> buyNonConsumable({required PurchaseParam purchaseParam}) async {
+    if (!(purchaseParam is AppStorePurchaseParam)) {
```

Comment:
Note here: this check is not necessary, the only thing missing from `PurchaseParam` is `simulatesAskToBuyInSandbox`, which is used only when testing. So we can actually make simulatesAskToBuyInSandbox defaults to `false`. so:
```dart
    await _skPaymentQueueWrapper.addPayment(SKPaymentWrapper(
        productIdentifier: purchaseParam.productDetails.id,
        quantity: 1,
        applicationUsername: purchaseParam.applicationUserName,
        simulatesAskToBuyInSandbox: (purchaseParam is AppStorePurchaseParam) ? purchaseParam.simulatesAskToBuyInSandbox: false
        requestData: null));
```
We can document it in the `AppStorePurchaseParam` how to test `simulatesAskToBuyInSandbox` in details, something like:

/// If you want to test `simulatesAskToBuyInSandbox`, you should create a `AppStorePurchaseParam` and use it in `buyNonConsumable` or `buyConsumable`.

Reply:
I really like this one, thanks for pointing it out. I have updated the code and the documentation of the `AppStorePurchaseParam` class.